libplatsupport/src/mach/eswin/timer.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2025, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <platsupport/mach/timer.h>
+#include <stdint.h>
+#include <utils/util.h>
+
+#define MAX_TIMEOUT_NS (ESWIN_TIMER_MAX_TICKS * (NS_IN_S / ESWIN_TIMER_TICKS_PER_SECOND))
+
+void eswin_timer_enable(eswin_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->regs->ctrl |= 0x1;
+}
+
+void eswin_timer_disable(eswin_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->regs->ctrl &= 0xfffffffe;
+}
+
+void eswin_timer_handle_irq(eswin_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    timer->value_h += 1;
+
+    /* Reading the EIO register acks the IRQ */
+    timer->regs->eoi;
+}
+
+uint64_t eswin_timer_get_time(eswin_timer_t *timer)
+{
+    assert(timer);
+    assert(timer->regs);
+
+    /* the timer value counts down from the load value */
+    uint64_t value_l = (uint64_t)(ESWIN_TIMER_MAX_TICKS - timer->regs->value);
+    uint64_t value_h = (uint64_t)timer->value_h;
+
+    /* Include unhandled interrupt in value_h */
+    if (timer->regs->int_status & 0x1) {
+        value_h += 1;
+    }
+
+    uint64_t value_ticks = (value_h << 32) | value_l;
+
+    /* convert from ticks to nanoseconds */
+    uint64_t value_whole_seconds = value_ticks / ESWIN_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ticks = value_ticks % ESWIN_TIMER_TICKS_PER_SECOND;
+    uint64_t value_subsecond_ns =
+        (value_subsecond_ticks * NS_IN_S) / ESWIN_TIMER_TICKS_PER_SECOND;
```

Comment:
One line please. If it doesn't fit, consider dropping the `value_` prefix.

Doing the calculation this way is more exact compared to what I did for the imx93 timer driver.